R/state.R

```diff
@@ -0,0 +1,288 @@
+#
+# @importFrom purrr partial
+#
+# @return a list, acc is the new state and counter
+#   as well as the action generated.
+gen.action <- function(commands, state, counter) {
+  possible <- Filter(function(command) {
+    !is.null ( command$gen(state) )
+  }, commands)
+  gen.with(gen.element( possible ), function (command) {
+    # The (symbolic) input for the command.
+    # Essentially this says which values it
+    # will read from.
+    gen.map(function(input) {
+      # Check the requires condition make sense.
+      # These requires functions are needed to
+      # ensure we have a good shrink.
+      require_ <- partial(command$require, state = state)
+      if (!do.call(require_, as.list(input)))
+        stop("Command generation arguments voilate requirements")
+
+      # Get a variable name we'll use for the output
+      # We just use sequential values, as we'll add
+      # them to the environment list here.
+      output   <- symbolic (counter)
+
+      # Build a new state which we can work with
+      update_  <- partial(command$update, state = state, output = output)
+      state_   <- do.call(update_ , as.list( input ))
+
+      # Build the action which can be run.
+      action_  <- structure(list(
+          title   = command$title
+        , input   = input
+        , output  = output
+        , execute = command$execute
+        , require = command$require
+        , update  = command$update
+        , ensure  = command$ensure
+      ), class = "action")
+
+      # Return the new symbolic state and the action to run.
+      list(acc = list(state = state_, counter = counter + 1), action = action_)
+    }, command$gen(state))
+  })
+}
+
+# Returns the actions and it's updates to the state
+# only if they're currently valid.
+check.valid  <- function(ok, state, action) {
+  require_  <- partial(action$require, state = state)
+  if (do.call(require_, as.list(action$input))) {
+    update_ <- partial(action$update, state = state, output = action$output)
+    state_  <- do.call(update_ , as.list(action$input))
+    list(ok = snoc(ok, action), state = state_)
+  } else {
+    list(ok = ok, state = state)
+  }
+}
+
+# After shrinking we may have an inconsistent state.
+# Run the state, ensuring the requirements are still
+# good.
+drop.invalid <- function(actions, initial.state) {
+  result <- Reduce(function(acc, action) {
+    check.valid( acc$ok, acc$state, action )
+  }, actions, init = list(ok = list(), state = initial.state))
+  result$ok
+}
+
+#' Generate a list of possible actions.
+#'
+#' @export
+#' @param initial.state the starting state to
+#'   build from which is appropriate for this
+#'   state machine generator.
+#' @param commands the list of commands which
+#'   we can select choose from. Only commands
+#'   appropriate for the state will actually be
+#'   selected.
+#'
+#' @importFrom purrr partial
+#'
+#' @return a list of actions to run during testing
+gen.actions <- function(initial.state, commands) {
```

Comment:
This gen could probably use some improvements. Its API should be more like `gen.list` and `gen.c`.